kernel/src/expressions/scalars.rs

```diff
@@ -39,6 +33,42 @@ impl ArrayData {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct MapData {
+    data_type: MapType,
+    pairs: Vec<(Scalar, Scalar)>,
+}
+
+impl MapData {
+    pub fn try_new(
+        data_type: MapType,
+        values: impl IntoIterator<Item = (impl Into<Scalar>, impl Into<Scalar>)>,
+    ) -> DeltaResult<Self> {
+        // TODO: check that inner types are primitive?
+        let pairs = values
+            .into_iter()
+            .map(|(k, v)| (k.into(), v.into()))
+            .collect();
+        Ok(Self { data_type, pairs })
+    }
+
+    // TODO: array.elements is deprecated? do we want to expose this? How will FFI get pairs for
+    // visiting?
+    pub fn pairs(&self) -> &[(Scalar, Scalar)] {
+        &self.pairs
+    }
+
+    pub fn map_type(&self) -> &MapType {
+        &self.data_type
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct StructData {
```

Comment:
I know this is old code, but if we're modeling this the same as `MapData` why not `Vec<(StructField, Scalar)>`?